include/alpaka/mem/buf/BufOacc.hpp

```diff
@@ -0,0 +1,652 @@
+                    }
+                    return pitchBytes;
+                }
+
+            public:
+                //-----------------------------------------------------------------------------
+                //! Constructor
+                template<
+                    typename TExtent>
+                ALPAKA_FN_HOST BufOaccImpl(
+                    DevOacc const & dev,
+                    TElem * const pMem,
+                    TExtent const & extent) :
+                        m_dev(dev),
+                        m_extentElements(extent::getExtentVecEnd<TDim>(extent)),
+                        m_pitchBytes(calculatePitchesFromExtents(m_extentElements)),
+                        m_pMem(pMem)
+                {
+                    ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
+
+                    static_assert(
+                        TDim::value == alpaka::Dim<TExtent>::value,
+                        "The dimensionality of TExtent and the dimensionality of the TDim template parameter have to be identical!");
+                    static_assert(
+                        std::is_same<TIdx, Idx<TExtent>>::value,
+                        "The idx type of TExtent and the TIdx template parameter have to be identical!");
+                }
+
+            public:
+                DevOacc m_dev;
+                Vec<TDim, TIdx> m_extentElements;
+                Vec<TDim, TIdx> m_pitchBytes;
+                TElem* m_pMem;
+
+                    BufOaccImpl(const BufOaccImpl&) = delete;
+                    BufOaccImpl(BufOaccImpl &&) = default;
+                BufOaccImpl& operator=(const BufOaccImpl&) = delete;
+                BufOaccImpl& operator=(BufOaccImpl&&) = default;
+                    ~BufOaccImpl()
+                {
+                    m_dev.makeCurrent();
+                    acc_free(m_pMem);
+                }
+            };
+        }
+    }
+
+    template<
+        typename TElem,
+        typename TDim,
+        typename TIdx>
+    class BufOacc
+    {
+    public:
+        //-----------------------------------------------------------------------------
+        //! Constructor
+        template<
+            typename TExtent>
+        ALPAKA_FN_HOST BufOacc(
+            DevOacc const & dev,
+            TElem * const pMem,
+            TExtent const & extent) :
+                m_spBufImpl(std::make_shared<oacc::detail::BufOaccImpl<TElem, TDim, TIdx>>(dev, pMem, extent))
+        {}
+
+            BufOacc(const BufOacc&) = default;
+            BufOacc(BufOacc &&) = default;
+        BufOacc& operator=(const BufOacc&) = default;
+        BufOacc& operator=(BufOacc&&) = default;
+
+        oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() {return *m_spBufImpl;}
+        const oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() const {return *m_spBufImpl;}
+
+        inline const Vec<TDim, TIdx>& extentElements() const {return m_spBufImpl->m_extentElements;}
```

Comment:
Class member functions are automatically inline. No need for the `inline` keyword.

Reply:
@jkelling